upgrades/upgrade_testing_framework/core/expectation.py

```diff
@@ -0,0 +1,96 @@
+from pathlib import Path
+import json
+from typing import List
+import operator
+
+from upgrade_testing_framework.core.versions_engine import EngineVersion
+
+class KnowledgeBaseDoesntExistException(Exception):
+    def __init__(self, knowledge_base_path):
+        super().__init__(f"The path specified for the knowledge base doesn't exist or is not a directory: {knowledge_base_path}")
+
+class ExpectationCantReadFileException(Exception):
+    def __init__(self, expecation_path, original_exception):
+        super().__init__(f"Unable to read test config file at path {expecation_path}.  Details: {str(original_exception)}")
+
+class ExpectationFileNotJSONException(Exception):
+    def __init__(self, expecation_path, original_exception):
+        super().__init__(f"The test config at path {expecation_path} is not parsible as JSON.  Details: {str(original_exception)}")
+
+class ExpectationMissingIdException(Exception):
+    def __init__(self):
+        super().__init__("An expectation was missing its id") # TODO make this better
+
+
+COMP_OPERATION = {
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "gte": operator.ge,
+    "lte": operator.le
+}
+
+
+class Expectation:
+    def __init__(self, raw_expectation: dict):
+        self.id = raw_expectation.get("id", None)
+        if self.id is None:
+            raise ExpectationMissingIdException()
+        self.description = raw_expectation.get("description", None)
+        self.version_filter = raw_expectation.get("versions", None)
+        
+
+    def is_relevant_to_version(self, version: EngineVersion) -> bool:
+        # If no version_filter, assume relevant to all versions
+        if self.version_filter is None:
+            return True
+
+        version_filters = self.version_filter if type(self.version_filter) is list else [self.version_filter]
+
+        # Version filters have sections which are ORed together, each of which has
```

Comment:
This for-loop is fairly complex; recommend adding comments breaking down what each part is doing to make it easier to maintain in the future.  Some suggestions added below:

```
        for section in version_filters: # For each AND'd set of conditionals
            section_valid = True

            # Check the individual conditionals within the AND'd set and exit this section if any yield False
            for op, comp_value in section.items():
                if not COMP_OPERATION[op](version, comp_value):
                    section_valid = False
                    break

            if section_valid: # All the AND'd conditionals yielded True
                return True # If any of the OR'd sections yields True, we return True
        return False # None of the OR'd sections yielded True
```